### ISSUE TITLE
Travis MacOS: linking error when building tests

### ISSUE THREAD

**Issue by wwylele** [citra-emu/citra#2183]
This issue comes from PR #2132, where adding some simple tests causes strange linking error, in MacOS build only. I did some experiment and here is the observation:

 - When tests is empty (as current master), no error.
 - When there is only a simple test that doesn't refer any other library (such as `REQUIRE(1 + 1 == 2);`), no error.
 - When there is a test referring function in core (_not video core_) (such as `REQUIRE(CoreTiming::GetClockFrequencyMHz() != 0);`), travis will error:
```
Undefined symbols for architecture x86_64:
  "_GLAD_GL_KHR_debug", referenced from:
      RendererOpenGL::Init() in libvideo_core.a(renderer_opengl.o)
  "_GLAD_GL_VERSION_3_3", referenced from:
      RendererOpenGL::Init() in libvideo_core.a(renderer_opengl.o)
  "_glad_glActiveTexture", referenced from:
      RendererOpenGL::SwapBuffers() in libvideo_core.a(renderer_opengl.o)
      RendererOpenGL::LoadColorToActiveGLTexture(unsigned char, unsigned char, unsigned char, TextureInfo const&) in libvideo_core.a(renderer_opengl.o)
      RendererOpenGL::ConfigureFramebufferTexture(TextureInfo&, GPU::Regs::FramebufferConfig const&) in libvideo_core.a(renderer_opengl.o)
      RendererOpenGL::LoadFBToScreenInfo(GPU::Regs::FramebufferConfig const&, ScreenInfo&) in libvideo_core.a(renderer_opengl.o)
      RendererOpenGL::DrawScreens() in libvideo_core.a(renderer_opengl.o)
      RendererOpenGL::InitOpenGLObjects() in libvideo_core.a(renderer_opengl.o)
      OpenGLState::Apply() const in libvideo_core.a(gl_state.o)
      ...
```
 - Adding `target_link_libraries(tests glad)` in `CMakeLists.txt` doesn't resolve the error above.
 - Adding some code that referring functions in glad such as `gladLoadGL();` _does_ resolve the error, even without `target_link_libraries(tests glad)`.

The experiment was done here: https://github.com/wwylele/citra/pull/9

**Comment by JayFoxRox** [comment]
Did you try these before: http://stackoverflow.com/questions/17470350/c-cmake-how-to-add-a-linker-flag-to-an-unused-library-when-the-library-is-sp ?

Also should we ask Apple about this? It feels like an error / bug in the linker where the dependencies for lib inclusion are only searched in object files. However, this is a theory, not something I actually confirmed (in any way):

*The video_core lib does depend on glad, but the linker doesn't check dependencies of libs. So you'll manually have to add that at the command line again.
* If you use video_core, but don't explicitly use glad in your own code, but still have glad on the lib list, it still won't link glad as it considers it unused.
* If you use video_core, but explicitly use glad in your own code and have glad on your lib list, it will be linked.
* If you don't use video_core glad is not used by any part of the application. As such, there is no error.

Reordering the libs on the command line did not help either.

**Comment by wwylele** [comment]
@JayFoxRox I tried adding `-Wl,--whole-archive` flag to the glad lib but on my gcc end it errors at multiple definitions:
```
/usr/lib/gcc/x86_64-linux-gnu/6/libgcc.a(_muldi3.o): In function `__multi3':
(.text+0x0): multiple definition of `__multi3'
/usr/lib/gcc/x86_64-linux-gnu/6/libgcc.a(_muldi3.o):(.text+0x0): first defined here
/usr/lib/gcc/x86_64-linux-gnu/6/libgcc.a(_negdi2.o): In function `__negti2':
(.text+0x0): multiple definition of `__negti2'
...
```

Also note that I didn't explicitly use video_core, but the core only. The reason video_core is linked in is probably that it is a dependency of the core (oh why the linker can recognize the dependency this time).

If this turns out to be a bug in the linker, then probably we need the work around (calling `gladLoadGL()`).

We need someone using Mac to reproduce the error on a simpler test case (i.e. without CMake and reduced structure).

**Comment by mailwl** [comment]
i am on mac, what exactly need to do?

**Comment by wwylele** [comment]
@mailwl Generally I want one who is more familiar to mac than me to figure out what is going on. Test whether #2132, or a master branch with a simple test `REQUIRE(CoreTiming::GetClockFrequencyMHz() != 0);`, can build the target `tests`, or it errors like Travis does. If it does error, try to reproduce the error in a simple C++ project that has a similar dependency structure as citra ( tests->core<->video_core->glad ) to confirm this is a bug in the linker. 
Well, I can't give more instructions since I don't know how it looks like on mac... we can discuss this on IRC or Discord.

**Comment by mailwl** [comment]
only `REQUIRE(CoreTiming::GetClockFrequencyMHz() != 0);` in sample.cpp: undefined ogl function errors
adding `gladLoadGL();` before - compiles successful, but running
`ctest -VV -C Release` lead to

> 
> 1: /Users/wl/projects/wwylele/citra/src/tests/sample.cpp:10: FAILED:
> 1: due to a fatal error condition:
> 1:   SIGSEGV - Segmentation violation signal

when i will be at home (in 2 hours), i may share TeamViewer to my mac

**Comment by MerryMage** [comment]
Should this be considered fixed by 282195b450?

**Comment by wwylele** [comment]
Yeah, although 282195b is a work-around, I think we can consider this fixed since it is more like a bug in clang. I will close this for now.
